app/services/interview_validations.rb

```diff
@@ -0,0 +1,93 @@
+class InterviewValidations
+  include ActiveModel::Model
+
+  APPLICATION_STATES_ALLOWING_CHANGES = %w[awaiting_provider_decision interviewing].freeze
+
+  attr_reader :interview, :today
+
+  delegate :provider, to: :interview
+  delegate :application_choice, to: :interview
+  delegate :current_course, to: :application_choice
+  delegate :date_and_time, to: :interview
+  delegate :location, to: :interview
+  delegate :additional_details, to: :interview
+  delegate :cancellation_reason, to: :interview
+
+  validates :date_and_time, presence: true
+  validates :application_choice, presence: true
+  validates :provider, presence: true
+  validates :location, presence: true, length: { maximum: 10240 }
+  validates :additional_details, length: { maximum: 10240 }
+  validates :cancellation_reason, length: { maximum: 10240 }
+
+  validate :require_training_or_ratifying_provider, if: -> { application_choice && interview.changed? }
+  validate :stop_changes_if_interview_already_passed, if: -> { interview.changed? }
+  validate :stop_changes_if_interview_already_cancelled, if: -> { interview.changed? }
+  validate :stop_changes_if_application_past_interviews_stage, if: -> { application_choice && interview.changed? }
+  validate :stop_cancellations_without_a_reason, if: -> { interview.changed? }
+  validate :updates_to_date_and_time, if: -> { interview.date_and_time_change }
+  validate :keep_date_and_time_before_rbd, if: -> { interview.date_and_time_change }
+
+  def initialize(interview:)
+    @interview = interview
+    @today = Time.zone.now.beginning_of_day
+  end
+
+  def rbd_date
+    @rbd_date ||= application_choice&.reject_by_default_at
+  end
+
+  def require_training_or_ratifying_provider
+    ratifying = current_course.accredited_provider
+    ratifying_provider_check = ratifying ? provider == ratifying : false
+
+    unless provider == current_course.provider || ratifying_provider_check
+      errors.add :provider, :training_or_ratifying_only
+    end
+  end
+
+  def stop_changes_if_interview_already_passed
+    if date_and_time && date_and_time < today
+      errors.add :base, :changing_a_past_interview
+    end
+  end
+
+  def stop_changes_if_interview_already_cancelled
+    old_cancelled_at = interview.cancelled_at_change&.first
```

Comment:
I'm finding this a little bit confusing. Why do we need to set the cancelled at before running the validations? Why not just check cancelled_at and then set the value inside the update service's transaction block?

Reply:
The intention here is that `IntervewValidations` is a one-stop/definitive set of rules concerning changes to interviews. Checking cancelled_at within the service (it'd be the cancel interview service in this case) is definitely possible but ideally services should only have to check if an interview is valid? or not, no further checks required.

The specific scenario addressed here is someone attempting to cancel the same interview a second time. `cancelled_at` will always carry the new value in this validation context, and we need to know if a previous value was set. If this is a new/first cancellation, `cancelled_at_change` will have `[nil, <new value>]`.

Reply:
Maybe we could use validation scopes instead, similar to the OfferWizard form object.

Reply:
I've modified `stop_changes_if_interview_already_cancelled` to use `cancelled_at_was`, hopefully it's a bit easier to read.

Reply:
After discussing this further, we have decided to extract the workflow-y validations out into a separate class, which does not care if all validation messages are preserved and raises errors, to be handled by the controller (e.g. changing a cancelled interview). These workflow-y validations are not really checking the value of a specific field but the state of the interview object or its application choice.

We have also introduced scopes for field-level validations (in `InterviewValidations`).